### DISCUSSION TITLE
Feature Request: more general batched SVD operations?

### DISCUSSION THREAD

**Discussion by LionSR** [jax-ml/jax#17609] [Ideas]
Hello Jax Developers and Community,

I've been delving into Jax's capabilities in Singular Value Decompositions (SVDs). While Jax provides support for `gesvdjBatched` (https://github.com/google/jax/pull/1314), I've observed that this implementation comes with a matrix size restriction (`m, n <= 32`). On looking further into cuSOLVER's offerings, I stumbled upon `gesvdaStridedBatched` which doesn't seem to be a part of Jax currently.

Here are some considerations related to `gesvdaStridedBatched`:

1. **General Matrix Support & No Size Restriction**: Unlike `gesvdjBatched`, `gesvdaStridedBatched` offers support for general matrices without the strict size limitations, potentially offering more flexibility.
  
2. **Usage in CuPy**: I noticed that CuPy has integrated `gesvdaStridedBatched` into their framework. Here's their [implementation](https://github.com/cupy/cupy/pull/3192). It's intriguing to see how they've leveraged it, and it made me wonder about its possible advantages if introduced in Jax.
   
3. **Performance Advantages**: An [NVIDIA presentation](https://developer.download.nvidia.com/video/gputechconf/gtc/2019/presentation/s9226-fast-singular-value-decomposition-on-gpus-v2.pdf) (Page 14) touches upon the speed-ups achievable with Batched GESVDA. It could be of interest to evaluate if these performance metrics align with what Jax aims for.

On a related note, it seems that AMD's equivalent, ROCm, does not have such matrix size restrictions, as seen [here](https://github.com/cupy/cupy/pull/4628). It does make me curious about the technical challenges or considerations behind these size restrictions.

I understand there are always multiple factors at play when deciding on integrations. It would be enlightening to hear the community's and developers' perspectives on this topic and whether `gesvdaStridedBatched` might find a place in Jax's roadmap.

Thank you for continually pushing the boundaries with Jax. Looking forward to learning more.

Edit 20230918: note that  gesvdaStridedBatched can be called in the pytorch interface with the commit https://github.com/pytorch/pytorch/commit/d136852bda4d9eb60927b7c8ce0f665686524643 of the pull request https://github.com/pytorch/pytorch/pull/74521.

Best wishes,
Ray

**Comment by adam-hartshorne** [comment]
This might be of interest,

Compositional Linear Algebra (CoLA)

CoLA is a framework for scalable linear algebra, automatically exploiting the structure often found in machine learning problems and beyond. CoLA supports both PyTorch and JAX.

https://github.com/wilson-labs/cola

**Comment by LionSR** [comment]
Pinging the author of https://github.com/google/jax/pull/1314  @hawkinsp :-) happy to help if there are some instructions (https://github.com/google/jax/pull/1099).

**Comment by TsingQAQ** [comment]
Hi JAX Community,

I wanted to follow up on this past discussion regarding SGESVDA (`gesvdaStridedBatched`) and check if it has/there are any plans for its integration into JAX. Given its ability to efficiently handle batched SVD without the size limitations of gesvdjBatched, it could be very useful.

One concrete motivation is randomized SVD, where gesvdaStridedBatched could significantly improve efficiency for low-rank approximations in large-scale settings (as one of the matrix dimensionality will typically be larger than 32). This is particularly important when working with large batch sizes, as current JAX implementations may face performance constraints.

If this is not currently a priority, are there any known workarounds in JAX? I checked the [CoLA library](https://github.com/wilson-labs/cola) (thanks to @adam-hartshorne  for sharing!), but it doesn’t seem to support advanced batch SVD.

Would love to hear any insights or thoughts on this!

Best,

**Reply by dfm** [reply]
Thanks for checking in! I'd say that it's not a team priority to add support for this algorithm, but it would be welcomed as a PR! Probably the most useful commit to look at is https://github.com/jax-ml/jax/commit/59e6e1c186504b78afe73d1f8b1fd533b640aef8, which adds all the GPU kernels for SVD. That combined with [the FFI tutorial](https://docs.jax.dev/en/latest/ffi.html) might be enough to get started on a PR, and I'd be very happy to provide feedback on a PR thread!
